File: test/separation_angle_test/FQ_separation_angle_test.py
import matplotlib
matplotlib.use('Agg')
import os
my_home = os.popen("echo $HOME").readlines()[0][:-1]
import numpy
from astropy.coordinates import SkyCoord
from astropy import units
import h5py
from sys import path, argv
path.append("E:/Github/astrophy-research/mylib/")
path.append('%s/work/mylib/' % my_home)
from plot_tool import Image_Plot
from subprocess import Popen


# generate 20000 galaxy pairs and calculate the separation angle
# for separation() debugging
cmd = argv[1]

if cmd == "generate":
    num = 10000
    num2 = int(num/2)
    ra = numpy.random.uniform(0, 360, num).reshape((num2,2))
    dec = numpy.random.uniform(-90, 90, num).reshape((num2,2))

    pts = numpy.zeros((num2, 6))
    pts[:,:2] = ra[:,:2]
    pts[:,2:4] = dec[:,:2]

    for i in range(num2):
        c1 = SkyCoord(ra=pts[i,0]*units.deg,dec=pts[i,2]*units.deg,frame ="fk5")
        c2 = SkyCoord(ra=pts[i,1]*units.deg,dec=pts[i,3]*units.deg,frame ="fk5")
        sep = c1.separation(c2)
        pts[i,4] = sep.radian
        pts[i,5] = sep.deg
        print("(%10.5f,%10.5f) <-- %10.5f rad (%10.5f deg) --> (%10.5f,%10.5f)"%(pts[i,0],pts[i,2],pts[i,4],pts[i,5],pts[i,1],pts[i,3]))

    h5f = h5py.File("./data/sep_test.hdf5","w")
    h5f["/data"] = pts
    h5f.close()

    cmd = "./FQ_sep_test 1 1 0 0 1"
    a = Popen(cmd, shell=True)
    a.wait()


if cmd == "compare":

    f = h5py.File("./data/sep_test_cpp.hdf5", "r")
    data1 = f["/data"].value
    f.close()

    f = h5py.File("./data/sep_test.hdf5", "r")
    data2 = f["/data"].value
    f.close()

    img = Image_Plot()
    img.subplots(1, 2)

    img.axs[0][0].hist(data1[:,0]-data2[:,4], 100)
    img.axs[0][1].hist(data1[:,1]-data2[:,5], 100)

    img.save_img("compare_result.png")
    # img.show_img()

File: test/separation_angle_test/FQ_separation_angle_test.cpp
#include<FQlib.h>

int main(int argc, char**argv)
{
	double RA1, RA2, DEC1, DEC2;
	double sep_angle;

	RA1 = atof(argv[1]);
	DEC1 = atof(argv[2]);
	RA2 = atof(argv[3]);
	DEC2 = atof(argv[4]);
	
	separation(RA1, DEC1, RA2, DEC2, sep_angle);
	
	char inform[200];
	sprintf(inform, "(%.4f, %.4f) <--  %.6f (%.4f deg) --> (%.4f, %.4f)", RA1, DEC1,  sep_angle, sep_angle / Pi * 180, RA2, DEC2);
	std::cout << inform << std::endl;

	if (argc > 5)
	{
		char path[200], set_name[20];
		int i, j, k, arr_size, num;
		int data_col = 6;

		sprintf(path, "./data/sep_test.hdf5");
		sprintf(set_name, "/data");

		read_h5_datasize(path, set_name, arr_size);
		num = arr_size / data_col;

		double *data = new double[arr_size];
		double *diff = new double[num * 2];
		double diff_theta;

		read_h5(path, set_name, data);

		for (i = 0; i < num; i++)
		{
			separation(data[i*data_col], data[i*data_col + 2], data[i*data_col + 1], data[i*data_col + 3], diff_theta);
			diff[i * 2] = diff_theta;
			diff[i * 2 + 1] = diff_theta / Pi * 180;
			if (fabs(diff_theta - data[i*data_col + 4]) > 0.001)
			{
				sprintf(inform, "P1: (%.4f, %.4f). P2: (%.4f, %.4f). Separation: %.6f (%.4f deg) %.6f",
					data[i*data_col], data[i*data_col + 2], data[i*data_col + 1], data[i*data_col + 3], diff[i * 2], diff[i * 2 + 1] / Pi * 180, data[i*data_col + 4]);

				std::cout << inform << std::endl;
			}
		}

		sprintf(path, "./data/sep_test_cpp.hdf5");
		sprintf(set_name, "/data");
		write_h5(path, set_name, diff, num, 2, TRUE);
	}
	return 0;
}